src/powerquery-formatter/themes/themes.ts

```diff
@@ -0,0 +1,389 @@
+        }
+    }
+}
+
+/**
+ * Theme tree element contains
+ */
+export class ThemeTrieElement {
+    constructor(
+        /**
+         * Current rule of the element
+         */
+        private readonly _mainRule: ThemeTrieElementRule,
+        /**
+         * Other rules of sharing the same scope name of the element but bear with parent scopes
+         */
+        private readonly _rulesWithParentScopes: ThemeTrieElementRule[] = [],
+        /**
+         * Children rules beneath current element
+         */
+        private readonly _children: Map<string, ThemeTrieElement> = new Map(),
+    ) {}
+
+    private static _sortBySpecificity(arr: ThemeTrieElementRule[]): ThemeTrieElementRule[] {
+        if (arr.length === 1) {
+            return arr;
+        }
+
+        arr.sort(this._cmpBySpecificity);
+
+        return arr;
+    }
+
+    private static _cmpBySpecificity(a: ThemeTrieElementRule, b: ThemeTrieElementRule): number {
+        if (a.scopeDepth === b.scopeDepth) {
+            const aParentScopes: string[] | undefined = a.parentScopes;
+            const bParentScopes: string[] | undefined = b.parentScopes;
+            const aParentScopesLen: number = !aParentScopes ? 0 : aParentScopes.length;
+            const bParentScopesLen: number = !bParentScopes ? 0 : bParentScopes.length;
+
+            if (aParentScopesLen === bParentScopesLen) {
+                // eslint-disable-next-line no-plusplus
+                for (let i: number = 0; i < aParentScopesLen; i++) {
+                    // aParentScopes and bParentScopes cannot be empty as aParentScopesLen is larger than 0
+                    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                    const aLen: number = aParentScopes![i].length;
+                    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                    const bLen: number = bParentScopes![i].length;
+
+                    if (aLen !== bLen) {
+                        return bLen - aLen;
+                    }
+                }
+            }
+
+            return bParentScopesLen - aParentScopesLen;
+        }
+
+        return b.scopeDepth - a.scopeDepth;
+    }
+
+    private getScopeHeadTailPair(scope: string): [string, string] {
+        const dotIndex: number = scope.indexOf(".");
+        let head: string;
+        let tail: string;
+
+        if (dotIndex === -1) {
+            head = scope;
+            tail = "";
+        } else {
+            head = scope.substring(0, dotIndex);
+            tail = scope.substring(dotIndex + 1);
+        }
+
+        return [head, tail];
+    }
+
+    public match(scope: string): ThemeTrieElementRule[] {
+        if (scope === "") {
+            // hit the tree rule of an empty scope name
+            return ThemeTrieElement._sortBySpecificity(
+                ([] as ThemeTrieElementRule[]).concat(this._mainRule).concat(this._rulesWithParentScopes),
```

Comment:
Can you replace this with `[this._mainRule].concat(this._rulesWithParentScopes)`?